frontend/src/pages/RoomPage.tsx

```diff
@@ -37,35 +68,135 @@ export const RoomPage = () => {
   }, []);
 
   useEffect(() => {
-    if (leaveRoom && socket) {
-      if (roomId) {
-        socket.emit(ROOM_EVENTS.LEAVE, roomId);
-      }
-      navigate("/dashboard");
+    if (socket && roomId) {
+      socket.emit(ROOM_EVENTS.JOIN, { roomId });
     }
-  }, [leaveRoom, socket, roomId, navigate]);
+  }, [socket, roomId]);
 
   useEffect(() => {
     if (!socket || !roomId) {
       return;
     }
 
-    socket.emit(ROOM_EVENTS.JOIN, roomId);
-
-    socket.on(ROOM_EVENTS.PARTNER_LEAVE, () => {
-      setMessage("The other user has left the room.");
+    socket.on(ROOM_EVENTS.DISCONNECT, (reason: string) => {
+      if (reason === SOCKET_IO_DISCONNECT_REASION.SERVER_CLOSE) {
+        navigate("/dashboard");
+        return;
+      }
+      setSelf((self) => ({ ...self, isConnected: false }));
+      enqueueSnackbar(`You have disconnected. Reconnecting...`, {
+        variant: "error",
+      });
     });
 
-    socket.on(ROOM_EVENTS.PARTNER_DISCONNECT, () => {
-      setMessage(
-        "The other user has disconnected. Waiting for reconnection...",
-      );
+    socket.on(ROOM_EVENTS.PARTNER_LEAVE, ({ userId }: PartnerLeavePayload) => {
+      if (userId === user?.userId) {
```

Comment:
If the self-disconnection is caught by the disconnect event, do we still need this guard conditional here?

Reply:
It's possible that the socket disconnects and reconnects immediately, and during the disconnection, the server fires the event to all participants including the user that caused the disconnection. But we should ignore such events if we caused it.